### DISCUSSION TITLE
RFC 1025- Database branches feedback

### DISCUSSION THREAD

**Discussion by raddevon** [geldata/rfcs#88] [General]
Please use this discussion to share your feedback on [RFC 1025- Database branches](https://github.com/edgedb/rfcs/blob/master/text/1025-branches.rst). In particular, we're interested in feedback on the CLI commands and the general CLI workflow proposed for this feature.

> Here's an example of how the rebase workflow is expected to work using "main" and "feature" branches:
>
> 1. Create a new "feature" VCS branch (a clone of the "main" branch) and a corresponding "feature" EdgeDB branch.
> 2. Work on the "feature" branch, add migrations, etc.
> 3. When it is time to merge the feature work back into the main branch we want to arrange things so that the "feature" branch is in a state that is a simple fast-forward w.r.t the "main" branch.
> 4. In order to achieve the above state we need to make sure "main" code branch as well as EdgeDB branch are both up-to-date.
> 5. Then we want to rebase the "feature" branch code on top of the "main" branch code.
> 6. After that we need to replicate the same rebase operation with the EdgeDB branch. Our CLI tools may need to first clone the "main" branch with the data into a "temp" branch. Then we can introspect the migration histories of "temp" and "feature" branches so that we can establish where they diverge. Take all the divergent migrations from the "feature" branch and apply them to the "temp" branch. If the operation is successful, drop the "feature" branch and rename "temp" to "feature". We now have successfully rebased "feature" branch on top of "main".
> 7. Since the state of "feature" is now a straightforward fast-forward w.r.t. the "main" branch we can finally merge "feature" back into main in VCS and then merge the EdgeDB branch as well (or rename "feature" EdgeDB branch into "main", if the old branch is no longer needed).
>
> In our CLI tools we need a edgedb branch rebase command to perform step 6) and also a edgedb branch merge command to perform a fast-forward merge (by copying the migration history and applying migrations).

**Comment by tomnz** [comment]
We would love this! Developers' local EdgeDB instance getting "out of sync" when working on a branch that touches the schema has been a continual pain point, especially if you need to jump around branches, abandon a branch, or try to rebase on someone elses' changes. The proposed CLI workflow makes a lot of sense!

**Comment by tdolsen** [comment]
This is awesome!

How about aligning the CLI commands more with Git? Instead of `edgedb branch switch` just use `edgedb checkout`. (And with a `-B` flag to create a new to match.)

It would be really nice if it were possible to see a log of migrations, with a message for each, just like `git log`. Then you'd need some way to attach a message to a migration, but that would be a welcome addition anyway.

**Comment by Arinelle** [comment]
Good news! I really like the idea of having branches to facilitate DataOps.

However, I think that reusing the `database` concept for the implementation is not such a good idea, as it will push you towards full data duplication instead of investigating ways to deduplicate it between branches.

I do not think that Git syntax is clear enough to warrant converging towards it. It is a major headache for most people to learn Git, even though they tend to forget that when they are used to it. A `switch` command is nice :-)
